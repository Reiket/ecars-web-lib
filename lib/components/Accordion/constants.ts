export const ACCORDION_TEST_ID = 'accordionTestId';
