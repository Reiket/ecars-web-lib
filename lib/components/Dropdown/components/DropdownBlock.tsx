import type {FC, ReactNode} from 'react';

interface Props {
  children: ReactNode;
}

export const DropdownBlock: FC<Props> = ({children}) => <div className="dropdown-menu__block">{children}</div>;
