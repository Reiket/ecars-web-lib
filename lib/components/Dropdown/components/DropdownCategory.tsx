import type {FC} from 'react';
import {Icons} from '@/services/icons';
import type {DropdownOption} from '@/components/Dropdown/constants';

export interface Props {
  item?: DropdownOption;
  handleOpen: () => void;
}

export const DropdownCategory: FC<Props> = ({item, handleOpen}) => (
  <button
    onClick={handleOpen}
    className="dropdown-menu__category"
  >
    {item?.label}
    {item?.value}
    <Icons.ArrowNarrowDown className="dropdown-menu__arrow" />
  </button>
);
