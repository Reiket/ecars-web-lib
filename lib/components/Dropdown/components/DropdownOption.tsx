import type {FC} from 'react';
import {cn} from '@/services/helpers';
import type {DropdownOption as DropdownOptionType} from '@/components/Dropdown/constants';

export interface Props {
  isSelected: boolean;
  option: DropdownOptionType;
  onSelect: (option: DropdownOptionType) => void;
}

export const DropdownOption: FC<Props> = ({option, isSelected, onSelect}) => {
  const {label, value, name} = option;
  const handleSelect = () => {
    onSelect(option);
  };
  return (
    <li
      data-value={value}
      onClick={handleSelect}
      className={cn('', 'dropdown-menu__option', {
        '_selected-option': isSelected,
      })}
    >
      {label && label} {name}
    </li>
  );
};
