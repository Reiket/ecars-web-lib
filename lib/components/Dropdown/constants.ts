import type {ReactNode} from 'react';

export interface DropdownOption {
  value: string;
  name: string;
  label?: ReactNode;
}

export interface DropdownProps {
  onSelect: (option: DropdownOption) => void;
  selectedOption: DropdownOption;
  options: DropdownOption[];
}

export const DROPDOWN_TEST_ID = 'dropdownTestId';

export const DROPDOWN_THEME = Object.freeze({
  DARK: 'dark',
  LIGHT: 'light',
});

export type DropdownThemeType = (typeof DROPDOWN_THEME)[keyof typeof DROPDOWN_THEME];
