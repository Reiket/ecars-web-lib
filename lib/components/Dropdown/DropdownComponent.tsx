import type {FC} from 'react';
import type {ElementProps} from '@/services/types';
import {cn} from '@/services/helpers';
import type {DropdownProps} from './constants';
import {DROPDOWN_TEST_ID} from './constants';
import {useClickOutside} from '@/services/hooks/useClickOutside';
import {Dropdown} from '@/components/Dropdown';

export interface Props extends ElementProps, DropdownProps {
  handleOpen: () => void;
  onClickOutside: (isOpen: boolean) => void;
  isOpen: boolean;
}
export const DropdownComponent: FC<Props> = ({
  category,
  block,
  options,
  onSelect,
  handleOpen,
  isOpen,
  onClickOutside,
}) => {
  const selectedOption = options.find((option) => option.value === category);
  const dropdownRef = useClickOutside(onClickOutside);
  return (
    <div
      data-testid={DROPDOWN_TEST_ID}
      ref={dropdownRef}
      className={cn(block, 'dropdown-menu', {
        '_dropdown-active': isOpen,
      })}
    >
      <Dropdown.Block>
        <Dropdown.Category
          handleOpen={handleOpen}
          item={selectedOption}
        />
      </Dropdown.Block>
      <Dropdown.List
        onSelect={onSelect}
        options={options}
        category={category}
      />
    </div>
  );
};
