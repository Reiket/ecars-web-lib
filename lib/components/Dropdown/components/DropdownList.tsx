import type {FC} from 'react';
import type {DropdownProps} from '@/components/Dropdown/constants';
import {Dropdown} from '@/components/Dropdown';

export const DropdownList: FC<DropdownProps> = ({selectedOption, options, onSelect}) => (
  <ul className="dropdown-menu__list">
    {options.map((option) => {
      const {value} = option;
      const isSelected = value === selectedOption.value;
      return (
        <Dropdown.Option
          key={value}
          option={option}
          onSelect={onSelect}
          isSelected={isSelected}
        />
      );
    })}
  </ul>
);
