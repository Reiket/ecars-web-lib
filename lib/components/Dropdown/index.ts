import {DropdownComponent} from '@/components/Dropdown/DropdownComponent';
import {DropdownList} from '@/components/Dropdown/components/DropdownList';
import {DropdownOption} from '@/components/Dropdown/components/DropdownOption';
import {DropdownCategory} from '@/components/Dropdown/components/DropdownCategory';
import {DropdownBlock} from '@/components/Dropdown/components/DropdownBlock';

export const Dropdown = Object.assign(DropdownComponent, {
  List: DropdownList,
  Option: DropdownOption,
  Category: DropdownCategory,
  Block: DropdownBlock,
});
