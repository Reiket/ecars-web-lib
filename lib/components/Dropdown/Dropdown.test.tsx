import {fireEvent, render, screen} from '@testing-library/react';
import {dropdownOptionsMock} from '@/services/mocks';
import {DROPDOWN_TEST_ID, DROPDOWN_THEME} from '@/components/Dropdown/constants';
import {Dropdown} from '@/components/Dropdown';

const defaultProps = {
  options: dropdownOptionsMock,
  onSelect: vi.fn(),
  handleOpen: vi.fn(),
  isOpen: false,
  onClickOutside: vi.fn(),
  selectedOption: {
    name: 'Option 1',
    value: '1',
  },
  theme: DROPDOWN_THEME.LIGHT,
};

describe('Dropdown Component', () => {
  const dropdownThemes = Object.values(DROPDOWN_THEME);
  dropdownThemes.forEach((theme) => {
    test(`renders the dropdown component correctly in ${theme} theme`, () => {
      const {container} = render(
        <Dropdown
          {...defaultProps}
          theme={theme}
        />,
      );
      const dropdownElement = screen.getByTestId(DROPDOWN_TEST_ID);
      expect(dropdownElement).toBeInTheDocument();
      expect(dropdownElement).toHaveClass(`dropdown-menu--${theme}`);
      expect(container).toMatchSnapshot();
    });
  });
  test('calls handleOpen when block is clicked', () => {
    const {container} = render(
      <Dropdown
        {...defaultProps}
        isOpen
      />,
    );
    const dropdownElement = screen.getByTestId(DROPDOWN_TEST_ID);
    const block = dropdownElement.querySelector('.dropdown-menu__category');
    if (block) {
      fireEvent.click(block);
    }
    expect(dropdownElement).toHaveClass('_dropdown-active');
    expect(defaultProps.handleOpen).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
  test('closes list on outside click', () => {
    render(<Dropdown {...defaultProps} />);
    fireEvent.mouseDown(document);
    expect(defaultProps.onClickOutside).toHaveBeenCalled();
  });
  test('highlights the selected option correctly', () => {
    const selectedOption = dropdownOptionsMock[0];
    const {container} = render(
      <Dropdown
        {...defaultProps}
        selectedOption={selectedOption}
        isOpen
      />,
    );
    const dropdownElement = screen.getByTestId(DROPDOWN_TEST_ID);
    const dropdownOptions = Array.from(dropdownElement.querySelectorAll('.dropdown-menu__option'));
    const activeOption = dropdownOptions.find((option) => option.classList.contains('_selected-option'));
    expect(activeOption).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  test('calls onSelect when an option is clicked', () => {
    render(
      <Dropdown
        {...defaultProps}
        isOpen
      />,
    );
    const dropdownElement = screen.getByTestId(DROPDOWN_TEST_ID);
    const options = Array.from(dropdownElement.querySelectorAll('.dropdown-menu__option'));
    const mockOption = dropdownOptionsMock[0];
    const firstMockOption = options.at(0);
    if (firstMockOption) {
      fireEvent.click(firstMockOption);
    }
    expect(defaultProps.onSelect).toHaveBeenCalledWith(mockOption);
  });
});
