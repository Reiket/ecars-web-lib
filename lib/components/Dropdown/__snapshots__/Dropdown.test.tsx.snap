// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dropdown Component > calls handleOpen when block is clicked 1`] = `
<div>
  <div
    class="dropdown-menu dropdown-menu--light _dropdown-active"
    data-testid="dropdownTestId"
  >
    <div
      class="dropdown-menu__block"
    >
      <button
        class="dropdown-menu__category"
        data-value="1"
      >
        Option 1
        <svg
          class="dropdown-menu__arrow"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
          />
        </svg>
      </button>
    </div>
    <ul
      class="dropdown-menu__list"
    >
      <li
        class="dropdown-menu__option _selected-option"
        data-value="1"
      >
         
        Option 1
      </li>
      <li
        class="dropdown-menu__option"
        data-value="2"
      >
         
        Option 2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Dropdown Component > highlights the selected option correctly 1`] = `
<div>
  <div
    class="dropdown-menu dropdown-menu--light _dropdown-active"
    data-testid="dropdownTestId"
  >
    <div
      class="dropdown-menu__block"
    >
      <button
        class="dropdown-menu__category"
        data-value="1"
      >
        Option 1
        <svg
          class="dropdown-menu__arrow"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
          />
        </svg>
      </button>
    </div>
    <ul
      class="dropdown-menu__list"
    >
      <li
        class="dropdown-menu__option _selected-option"
        data-value="1"
      >
         
        Option 1
      </li>
      <li
        class="dropdown-menu__option"
        data-value="2"
      >
         
        Option 2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Dropdown Component > renders the dropdown component correctly in dark theme 1`] = `
<div>
  <div
    class="dropdown-menu dropdown-menu--dark"
    data-testid="dropdownTestId"
  >
    <div
      class="dropdown-menu__block"
    >
      <button
        class="dropdown-menu__category"
        data-value="1"
      >
        Option 1
        <svg
          class="dropdown-menu__arrow"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
          />
        </svg>
      </button>
    </div>
    <ul
      class="dropdown-menu__list"
    >
      <li
        class="dropdown-menu__option _selected-option"
        data-value="1"
      >
         
        Option 1
      </li>
      <li
        class="dropdown-menu__option"
        data-value="2"
      >
         
        Option 2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Dropdown Component > renders the dropdown component correctly in light theme 1`] = `
<div>
  <div
    class="dropdown-menu dropdown-menu--light"
    data-testid="dropdownTestId"
  >
    <div
      class="dropdown-menu__block"
    >
      <button
        class="dropdown-menu__category"
        data-value="1"
      >
        Option 1
        <svg
          class="dropdown-menu__arrow"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
          />
        </svg>
      </button>
    </div>
    <ul
      class="dropdown-menu__list"
    >
      <li
        class="dropdown-menu__option _selected-option"
        data-value="1"
      >
         
        Option 1
      </li>
      <li
        class="dropdown-menu__option"
        data-value="2"
      >
         
        Option 2
      </li>
    </ul>
  </div>
</div>
`;
